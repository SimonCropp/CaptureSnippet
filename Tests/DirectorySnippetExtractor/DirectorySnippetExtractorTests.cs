using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using ApprovalTests.Reporters;
using CaptureSnippets;
using NuGet.Versioning;
using NUnit.Framework;
using ObjectApproval;

[TestFixture]
[UseReporter(typeof(AllFailingTestsClipboardReporter), typeof(DiffReporter))]
public class DirectorySnippetExtractorTests
{
    [Test]
    public void VerifyLambdasAreCalled()
    {
        var versionAndPaths = new ConcurrentBag<CapturedVersionAndPath>();
        var directories = new ConcurrentBag<CapturedDirectory>();
        var files = new ConcurrentBag<CapturedFile>();
        var translatePackages = new ConcurrentBag<CapturedTranslatePackage>();
        var targetDirectory = Path.Combine(TestContext.CurrentContext.TestDirectory, "DirectorySnippetExtractor");
        var data = PathData.With(VersionRange.All, "package", Component.Undefined);
        var result = new TestResult();
        var extractor = new DirectorySnippetExtractor(
            extractPathData: path =>
            {
                var versionAndPath = new CapturedVersionAndPath
                {
                    Path = path
                };
                versionAndPaths.Add(versionAndPath);
                return data;
            },
            directoryFilter: path =>
            {
                directories.Add(new CapturedDirectory {Path = path});
                return true;
            },
            fileFilter: path =>
            {
                files.Add(new CapturedFile {Path = path});
                return true;
            },
            translatePackage: alias =>
            {
                translatePackages.Add(new CapturedTranslatePackage {Alias = alias});
                return alias;
            }
        );
        extractor.FromDirectory(targetDirectory, VersionRange.None, Package.Undefined, Component.Undefined)
            .GetAwaiter()
            .GetResult();
        result.Files = files.OrderBy(file => file.Path).ToList();
        result.Directories = directories.OrderBy(file => file.Path).ToList();
        result.VersionAndPaths = versionAndPaths.OrderBy(file => file.Path).ToList();
        result.TranslatePackages = translatePackages.OrderBy(file => file.Alias).ToList();
        ObjectApprover.VerifyWithJson(result, s => s.Replace(@"\\", @"\").Replace(targetDirectory, @"root\"));
    }

    public class CapturedTranslatePackage
    {
        public string Alias;
    }

    public class TestResult
    {
        public List<CapturedVersionAndPath> VersionAndPaths;
        public List<CapturedDirectory> Directories;
        public List<CapturedFile> Files;
        public List<CapturedTranslatePackage> TranslatePackages;
    }

    public class CapturedVersionAndPath
    {
        public string Path;
    }

    public class CapturedDirectory
    {
        public string Path;
    }

    public class CapturedFile
    {
        public string Path;
    }

}