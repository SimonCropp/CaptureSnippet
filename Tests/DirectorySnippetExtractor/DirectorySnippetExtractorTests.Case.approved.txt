{
  "Components": [
    {
      "Identifier": "componenty",
      "Packages": [
        {
          "Identifier": "Packagea",
          "Versions": [
            [
              {
                "Key": "snippet",
                "Language": "txt",
                "Package": "Packagea",
                "Version": ">= 2.0.0",
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
                "IsInError": false
              },
              {
                "Key": "componentsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
                "IsInError": false
              },
              {
                "Key": "globalsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
                "IsInError": false
              }
            ],
            [
              {
                "Key": "snippet",
                "Language": "txt",
                "Package": "packageA",
                "Version": ">= 1.0.0 && < 2.0.0",
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
                "IsInError": false
              },
              {
                "Key": "componentsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
                "IsInError": false
              },
              {
                "Key": "globalsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
                "IsInError": false
              }
            ]
          ],
          "Snippets": [
            {
              "Key": "snippet",
              "Language": "txt",
              "Package": "Packagea",
              "Version": ">= 2.0.0",
              "Value": "Some code",
              "Error": null,
              "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
              "IsInError": false
            },
            {
              "Key": "componentsharedsnippet",
              "Language": "txt",
              "Package": null,
              "Version": null,
              "Value": "Some code",
              "Error": null,
              "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
              "IsInError": false
            },
            {
              "Key": "globalsharedsnippet",
              "Language": "txt",
              "Package": null,
              "Version": null,
              "Value": "Some code",
              "Error": null,
              "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
              "IsInError": false
            },
            {
              "Key": "snippet",
              "Language": "txt",
              "Package": "packageA",
              "Version": ">= 1.0.0 && < 2.0.0",
              "Value": "Some code",
              "Error": null,
              "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
              "IsInError": false
            }
          ],
          "Lookup": {
            "snippet": [
              {
                "Key": "snippet",
                "Language": "txt",
                "Package": "Packagea",
                "Version": ">= 2.0.0",
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
                "IsInError": false
              },
              {
                "Key": "snippet",
                "Language": "txt",
                "Package": "packageA",
                "Version": ">= 1.0.0 && < 2.0.0",
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
                "IsInError": false
              }
            ],
            "componentsharedsnippet": [
              {
                "Key": "componentsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
                "IsInError": false
              }
            ],
            "globalsharedsnippet": [
              {
                "Key": "globalsharedsnippet",
                "Language": "txt",
                "Package": null,
                "Version": null,
                "Value": "Some code",
                "Error": null,
                "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
                "IsInError": false
              }
            ]
          }
        }
      ],
      "Directory": "root\\DirectorySnippetExtractor/Case\componenty",
      "Snippets": [
        {
          "Key": "snippet",
          "Language": "txt",
          "Package": "Packagea",
          "Version": ">= 2.0.0",
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
          "IsInError": false
        },
        {
          "Key": "componentsharedsnippet",
          "Language": "txt",
          "Package": null,
          "Version": null,
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
          "IsInError": false
        },
        {
          "Key": "globalsharedsnippet",
          "Language": "txt",
          "Package": null,
          "Version": null,
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
          "IsInError": false
        },
        {
          "Key": "snippet",
          "Language": "txt",
          "Package": "packageA",
          "Version": ">= 1.0.0 && < 2.0.0",
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
          "IsInError": false
        }
      ],
      "Shared": [
        {
          "Key": "globalsharedsnippet",
          "Language": "txt",
          "Package": null,
          "Version": null,
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
          "IsInError": false
        },
        {
          "Key": "componentsharedsnippet",
          "Language": "txt",
          "Package": null,
          "Version": null,
          "Value": "Some code",
          "Error": null,
          "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
          "IsInError": false
        }
      ],
      "AllVersions": [
        [
          {
            "Key": "snippet",
            "Language": "txt",
            "Package": "Packagea",
            "Version": ">= 2.0.0",
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
            "IsInError": false
          },
          {
            "Key": "componentsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
            "IsInError": false
          },
          {
            "Key": "globalsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
            "IsInError": false
          }
        ],
        [
          {
            "Key": "snippet",
            "Language": "txt",
            "Package": "packageA",
            "Version": ">= 1.0.0 && < 2.0.0",
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
            "IsInError": false
          },
          {
            "Key": "componentsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
            "IsInError": false
          },
          {
            "Key": "globalsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
            "IsInError": false
          }
        ]
      ],
      "Lookup": {
        "snippet": [
          {
            "Key": "snippet",
            "Language": "txt",
            "Package": "Packagea",
            "Version": ">= 2.0.0",
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
            "IsInError": false
          },
          {
            "Key": "snippet",
            "Language": "txt",
            "Package": "packageA",
            "Version": ">= 1.0.0 && < 2.0.0",
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
            "IsInError": false
          }
        ],
        "componentsharedsnippet": [
          {
            "Key": "componentsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
            "IsInError": false
          }
        ],
        "globalsharedsnippet": [
          {
            "Key": "globalsharedsnippet",
            "Language": "txt",
            "Package": null,
            "Version": null,
            "Value": "Some code",
            "Error": null,
            "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
            "IsInError": false
          }
        ]
      }
    }
  ],
  "Directory": "root\\DirectorySnippetExtractor/Case",
  "AllSnippets": [
    {
      "Key": "snippet",
      "Language": "txt",
      "Package": "Packagea",
      "Version": ">= 2.0.0",
      "Value": "Some code",
      "Error": null,
      "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
      "IsInError": false
    },
    {
      "Key": "componentsharedsnippet",
      "Language": "txt",
      "Package": null,
      "Version": null,
      "Value": "Some code",
      "Error": null,
      "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
      "IsInError": false
    },
    {
      "Key": "globalsharedsnippet",
      "Language": "txt",
      "Package": null,
      "Version": null,
      "Value": "Some code",
      "Error": null,
      "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
      "IsInError": false
    },
    {
      "Key": "snippet",
      "Language": "txt",
      "Package": "packageA",
      "Version": ">= 1.0.0 && < 2.0.0",
      "Value": "Some code",
      "Error": null,
      "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
      "IsInError": false
    }
  ],
  "Shared": [
    {
      "Key": "globalsharedsnippet",
      "Language": "txt",
      "Package": null,
      "Version": null,
      "Value": "Some code",
      "Error": null,
      "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
      "IsInError": false
    }
  ],
  "SnippetsInError": [],
  "Lookup": {
    "snippet": [
      {
        "Key": "snippet",
        "Language": "txt",
        "Package": "Packagea",
        "Version": ">= 2.0.0",
        "Value": "Some code",
        "Error": null,
        "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Packagea_2\code.txt(2-3)",
        "IsInError": false
      },
      {
        "Key": "snippet",
        "Language": "txt",
        "Package": "packageA",
        "Version": ">= 1.0.0 && < 2.0.0",
        "Value": "Some code",
        "Error": null,
        "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\packageA_1\code.txt(2-3)",
        "IsInError": false
      }
    ],
    "componentsharedsnippet": [
      {
        "Key": "componentsharedsnippet",
        "Language": "txt",
        "Package": null,
        "Version": null,
        "Value": "Some code",
        "Error": null,
        "FileLocation": "root\\DirectorySnippetExtractor/Case\componenty\Shared\code.txt(2-3)",
        "IsInError": false
      }
    ],
    "globalsharedsnippet": [
      {
        "Key": "globalsharedsnippet",
        "Language": "txt",
        "Package": null,
        "Version": null,
        "Value": "Some code",
        "Error": null,
        "FileLocation": "root\\DirectorySnippetExtractor/Case\Shared\code.txt(2-3)",
        "IsInError": false
      }
    ]
  }
}